File: leet-code/75. Sort Colors/SortColors75_2.java
package com.example;

public class SortColors75_2 {
    public void sortColors(int[] nums) {
        int red = 0;
        int white = 0;
        int blue = nums.length - 1;

        while (white <= blue) {
            if (nums[white] < 1) {
                swap(nums, red, white);
                red += 1;
                white += 1;
            } else if (nums[white] > 1 ) {
                swap(nums, white, blue);
                blue -= 1;
            } else {
                white += 1;
            }
        }
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: leet-code/75. Sort Colors/SortColors75_2.kt
package com.example

class SortColors75_2 {
    fun sortColors(nums: IntArray): Unit {
        var i = 0
        var j = 0
        var k = nums.size - 1
        while (j <= k) {
            if (nums[j] < 1) {
                swap(nums, i, j)
                i++
                j++
            } else if (nums[j] > 1) {
                swap(nums, j, k)
                k--
            } else {
                j++
            }
        }
    }

    private fun swap(nums: IntArray, idx1: Int, idx2: Int) {
        val temp = nums[idx1]
        nums[idx1] = nums[idx2]
        nums[idx2] = temp
    }
}
